inference/primitive.go

```diff
@@ -33,72 +37,212 @@ import (
+// correct, but not others). This leads to mismatches in our inference engine: we cannot find
+// nilabilities of upstream sites in the imported InferredMap since the positions of the primitive
+// sites are different. The primitivizer here contains logic to fix such discrepancies.
+// Specifically, local incorrect upstream position will be fixed to match the upstream sites in the
+// inferred map (such that all position information in primitiveSite is always correct).
+//
+// [archive importer]: https://github.com/golang/tools/blob/fa12f34b4218307705bf0365ab7df7c119b3653a/internal/gcimporter/bimport.go#L59-L69
+type primitivizer struct {
+	pass *analysis.Pass
+	// upstreamObjPositions maps "<pkg path>.<object path>" to the correct position.
+	upstreamObjPositions map[string]token.Position
+	// curDir is the current working directory, which is used to trim the prefix (e.g., bazel
+	// random sandbox prefix) from the file names for cross-package references.
+	curDir string
+}
+
+// newPrimitivizer returns a new and properly-initialized primitivizer.
+func newPrimitivizer(pass *analysis.Pass) *primitivizer {
+	// To tackle the position discrepancies for upstream sites, we have added an ObjectPath field
+	// to primitiveSite, which can be used to uniquely identify an exported object relative to the
+	// package. Then, we simply cache the correct position information when importing
+	// InferredMaps, since the positions collected in the upstream analysis are always correct.
+	// Later when querying upstream objects in downstream analysis, we can look up the cache and
+	// fill in the correct position in the returned primitive site instead.
+
+	// Create a cache for upstream object positions.
+	upstreamObjPositions := make(map[string]token.Position)
+	for _, pkgFact := range pass.AllPackageFacts() {
+		importedMap, ok := pkgFact.Fact.(*InferredMap)
+		if !ok {
+			continue
+		}
+
+		importedMap.Range(func(site primitiveSite, _ InferredVal) bool {
+			if site.ObjectPath == "" {
+				return true
+			}
+
+			objRepr := site.PkgPath + "." + string(site.ObjectPath)
+			if existing, ok := upstreamObjPositions[objRepr]; ok && existing != site.Position {
+				panic(fmt.Sprintf(
+					"conflicting position information on upstream object %q: existing: %v, got: %v",
+					objRepr, existing, site.Position,
+				))
+			}
+			upstreamObjPositions[objRepr] = site.Position
+			return true
+		})
+	}
+
+	// Find the current working directory (e.g., random sandbox prefix) for trimming the file names.
+	cwd, err := os.Getwd()
+	if err != nil {
+		panic(fmt.Sprintf("cannot get current working directory: %v", err))
+	}
+
+	return &primitivizer{
+		pass:                 pass,
+		upstreamObjPositions: upstreamObjPositions,
+		curDir:               cwd,
+	}
+}
+
+// fullTrigger returns the primitive version of the full trigger.
+func (p *primitivizer) fullTrigger(trigger annotation.FullTrigger) primitiveFullTrigger {
```

Comment:
Could we rename `fullTrigger()` to something more intuitive (e.g., `simplifiedFullTrigger()`)? I think it could be a little confusing to call this `fullTrigger`, when we are actually returning its simplified or primitive version.

Reply:
I see your point. I would still prefer to rename the method to make it more intuitive and decouple from the `full trigger` connotation  that exists otherwise in the code. But this is only a small nit, and can be ignored :)

Reply:
I see. Somehow I still prefer the current name given that `primitive.simplifiedFullTrigger` reads a little confusing, `primitive.primitiveFullTrigger` is stuttering, and `primitive.ToPrimitiveFullTrigger` is a bit unnecessarily long.

But I get your point! Since this `primitivizer` struct and logic is only within `inference` package (unexported), and there is no immediate plan to export it. I feel it's relatively safe to keep it this way. If we ever need to export it in the future, we can make it clearer by adding some prefixes 😃 

Hope that also makes sense to you